Sticky Selection

Templarfreak opened the pull request:
Refine micro mechanics. Allow for more organic feeling controls.

This mechanic is very helpful for a number of micro techniques, such as:

Splitting.
Concaves.
And Kiting.

Mechanics that allow for players to show off their skill in unit control
makes a more exciting to watch and more rewarding to play game without raising the skill floor.

Video showcasing the change in work:

http://www.mediafire.com/watch/dj9i9bba4r4inbt/OpenRA.Game_2015-05-17_03-20-27-896.avi

Commit:
Sticky Selection: Refine micro mechanics. Allow for more organic feeling controls.

This mechanic is very helpful for a number of micro techniques, such as:

Splitting.
Concaves.
And Kiting.

Mechanics that allow for players to show off their skill in unit control
makes a more exciting to watch and more rewarding to play game.

Templarfreak commented:
Er, to clarify what this does is if someone makes a drag select box then it won't deselect units that are currently in the player's selection if there are no units to select in that drag select box.

x-a-n-a-x commented:
a GIF or small video would be cool to clearly see what it changes.

abcdefg30 commented:
Could you please use tabs instead of spaces? Thanks.

Templarfreak commented:
Spaces happened because I copy/pasted it over from another file. I used tabs originally but it turned them into spaces. xD

Will make a video.

Templarfreak commented:
Here is a video: http://www.mediafire.com/watch/dj9i9bba4r4inbt/OpenRA.Game_2015-05-17_03-20-27-896.avi

WolfGaming commented:
If you could make it an option that you can switch between the two behaviors I would be very happy since I loath playing with sticky selection xD

Templarfreak commented:
I just finished making it so Sticky Selection is only on when Left-Click orders is on. For a proper toggle, I'll have to spend tomorrow figuring out how the UI works for that.

Commit:
Hotfix:

Made it so Sticky Selection is only on when Left-Click Orders are on.
Applied the sound bug fix.

Commit:
Hotfix:

Sticky Select is now only on when Left-Click Orders are on. (Temp solution, can add
an actual setting later)
Applied sound bug fix.

Templarfreak commented:
I have no idea how to "Squish" all those changes into one...

Unit158 commented:
```bash
git fetch upstream bleed # Or "openra bleed", whichever you use.
git rebase -i FETCH_HEAD
#now, choose the commits you'd like to squash
```

Templarfreak commented:
For both of those it says it's not a git repository.

Unit158 commented:
You'll need to set OpenRA/OpenRA as a remote then. Come on the IRC for help.

dan9550 commented:
What @WolfGaming said if this was by default you would make me a sad panda lol

Toggle-able in game would be my preference.

Mailaender commented:
This seems to be very useful in terms of improving usability. However we can't accept this GitHub web interface broken commit spam. https://help.github.com/articles/using-git-rebase/ has a tutorial, but as this is often confusing for beginners, getting talked through the process in the chat may be necessary. See http://www.openra.net/community/ on how to reach us.

Templarfreak commented:
Changes requested made. (I think)

Templarfreak commented:
Changed comment to look not dumb.

Mailaender commented on OpenRA.Game/Selection.cs:
```diff
@@ -33,40 +33,65 @@ public bool Contains(Actor a)
+                    sel.Selected(a);
+            foreach (var ns in world.WorldActor.TraitsImplementing<INotifySelection>())
+                ns.SelectionChanged();
+        }
```
Can you replace the space indention with tabs please. We provide an @EditorConfig that may be of help to you.

Templarfreak commented:
I don't know how to use that... What exactly do I do to compile it? I don't see a project file.

Mailaender commented on OpenRA.Game/Selection.cs:
```diff
@@ -33,40 +33,65 @@ public bool Contains(Actor a)
+            {
+            if (voicedUnit != null)
+                Sound.PlayVoice("Select", voicedUnit, voicedUnit.Owner.Country.Race);
+                                           }
```
Something went wrong here, too with indentions.

Mailaender commented:
See https://msdn.microsoft.com/en-us/library/ms165330%28v=vs.90%29.aspx?f=255&MSPPError=-2147217396

Mailaender commented:
See http://edglogowski.azurewebsites.net/?p=10841

Mailaender commented on OpenRA.Game/Selection.cs:
```diff
@@ -33,40 +33,65 @@ public bool Contains(Actor a)
 		}
 
-		public void Combine(World world, IEnumerable<Actor> newSelection, bool isCombine, bool isClick)
+	public void Combine(World world, IEnumerable<Actor> newSelection, bool isCombine, bool isClick)
```
Everything is off by one tab/indention, but the spaces are gone.

Mailaender commented on OpenRA.Game/Selection.cs:
```diff
@@ -35,37 +35,62 @@ public bool Contains(Actor a)
-				if (isCombine)
+			if (isCombine)
+				actors.UnionWith(newSelection);
+			 else
```
spurious space

Mailaender commented on OpenRA.Game/Selection.cs:
```diff
@@ -35,37 +35,62 @@ public bool Contains(Actor a)
 		public void Combine(World world, IEnumerable<Actor> newSelection, bool isCombine, bool isClick)
 		{
-			if (isClick)
+		var classicmouse = Game.Settings.Game.UseClassicMouseStyle;
```
Still missing one indention step.

Mailaender commented on OpenRA.Game/Selection.cs:
```diff
@@ -35,37 +35,62 @@ public bool Contains(Actor a)
-					actors.SymmetricExceptWith(adjNewSelection);
-				else
+				if (classicmouse == true)
+			{
```
This is off by two indentions.

Templarfreak commented:
Okay, if all the indentation's right now I'll go ahead and squash it all again.

Templarfreak commented:
Okay, gonna squish 'em now.

Commit:
Attempting to add checkbox to turn off sticky selection instead of it relying
on Left-Click Orders.

Commit:
Testing stuff

obrakmann commented:
* A few codestyle issues. At a first glance, the indentation is wrong in some lines.
* The checkbox will need to be added to the cnc mod as well
* You never actually use Settings.StickySelection 
* We'll have to figure out the best place for the checkbox to live in the UI, because it's so cramped in there and everything collides

obrakmann commented on OpenRA.Game/Selection.cs:
```diff
@@ -42,8 +43,19 @@ public void Combine(World world, IEnumerable<Actor> newSelection, bool isCombine
+					else
 					actors.Clear();
-					actors.UnionWith(adjNewSelection);
+					actors.UnionWith(newSelection);
```
This block needs one more level of indentation, and it has to be wrapped in braces, too.

obrakmann commented on OpenRA.Game/Selection.cs:
```diff
@@ -52,14 +64,26 @@ public void Combine(World world, IEnumerable<Actor> newSelection, bool isCombine
+							actors.UnionWith(newSelection);
+						else
+						actors.Clear();
+						actors.UnionWith(newSelection);
```
Here as well: one more level of indentation, and braces are missing on the else block

obrakmann commented on OpenRA.Game/Selection.cs:
```diff
@@ -52,14 +64,26 @@ public void Combine(World world, IEnumerable<Actor> newSelection, bool isCombine
+					}
+					else
 					actors.Clear();
 					actors.UnionWith(newSelection);
```
and again

Templarfreak commented:
I really have no idea how many times I've set those indentations and VS undid them...

Commit:
Working on adding a button dedicated to toggling Sticky Selection.

WolfGaming commented:
You probably havent set it up to keep tabs or something.

Try going to Options/Text Editor/C#/Tabs and check if the tab/indent size are both set to 4 and Keep tabs is selected.

abcdefg30 commented:
Needs a rebase.

Templarfreak commented:
Will need to come back to this later, lots of stuff just came up. Sorry, guys! (Don't worry, I'm not disappearing. I can come back to this in a couple weeks)

pchote commented:
Sure thing @Templarfreak.  I'll close this pr for now, and you can either either reopen it if/when you decide to continue on this version, or open a new one if you decide to do it over.
